onnxruntime/featurizers_ops/cpu/mode_imputer_transformer.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/common.h"
+#include "core/framework/data_types.h"
+#include "core/framework/data_types_internal.h"
+#include "core/framework/op_kernel.h"
+
+#include "Featurizers/ModeImputerFeaturizer.h"
+#include "Featurizers/../Archive.h"
+
+namespace onnxruntime {
+namespace featurizers {
+
+template <typename T>
+struct OutputTypeMapper {};
+template <>
+struct OutputTypeMapper<float> { using type = float; };
+template <>
+struct OutputTypeMapper<double> { using type = double; };
+template <>
+struct OutputTypeMapper<std::string> { using type = std::string; };
+
+inline float const& PreprocessOptional(float const& value) { return value; }
+inline double const& PreprocessOptional(double const& value) { return value; }
+inline nonstd::optional<std::string> PreprocessOptional(std::string value) {
+  return value.empty() ? nonstd::optional<std::string>() : nonstd::optional<std::string>(std::move(value));
+}
+
+// Hack
+namespace FeatDetails = Microsoft::Featurizer::Featurizers::Details;
+
+template <typename T>
+struct ModeImputerTransformerT : public FeatDetails::ModeImputerEstimatorImpl<T, std::numeric_limits<size_t>::max()>::TransformerType {
+  using BaseType = typename FeatDetails::ModeImputerEstimatorImpl<T, std::numeric_limits<size_t>::max()>::TransformerType;
+  explicit ModeImputerTransformerT(Microsoft::Featurizer::Archive& ar) : BaseType(ar) {}
+};
+
+template <typename T>
+struct ModeImputerTransformerImpl {
+  void operator()(OpKernelContext* ctx) const {
+    // Create the transformer
+    ModeImputerTransformerT<T> transformer(
+        [ctx]() {
+          const auto* state_tensor(ctx->Input<Tensor>(0));
+          const uint8_t* const state_data(state_tensor->Data<uint8_t>());
+
+          Microsoft::Featurizer::Archive archive(state_data, state_tensor->Shape().GetDims()[0]);
+          return ModeImputerTransformerT<T>(archive);
+        }());
+
+    // Get the input
+    const auto* input_tensor(ctx->Input<Tensor>(1));
+    const T* input_data(input_tensor->Data<T>());
+
+    // Prepare the output
+    Tensor* output_tensor(ctx->Output(0, input_tensor->Shape()));
+    typename OutputTypeMapper<T>::type* output_data(output_tensor->MutableData<typename OutputTypeMapper<T>::type>());
+
+    // Execute
+    const int64_t length(input_tensor->Shape().Size());
+
+    for (int64_t i = 0; i < length; ++i) {
+      output_data[i] = transformer.execute(PreprocessOptional(input_data[i]));
+    }
+  }
+};
+
+class ModeImputerTransformer final : public OpKernel {
```

Comment:
Do you want to allow copy/move operations for this class?

Reply:
The class is stateless and kernels are never copied. Not sure I understood the question.

---
In reply to: [373740339](https://github.com/microsoft/onnxruntime/pull/2958#discussion_r373740339) [](ancestors = 373740339)